server/src/main/java/org/opensearch/action/support/clustermanager/TransportClusterManagerNodeAction.java

```diff
@@ -134,32 +136,51 @@ protected boolean localExecute(Request request) {
 
     @Override
     protected void doExecute(Task task, final Request request, ActionListener<Response> listener) {
-        ClusterState state = clusterService.state();
-        logger.trace("starting processing request [{}] with cluster state version [{}]", request, state.version());
         if (task != null) {
             request.setParentTask(clusterService.localNode().getId(), task.getId());
         }
-        new AsyncSingleAction(task, request, listener).doStart(state);
+        new AsyncSingleAction(task, request, listener).run();
     }
 
     /**
      * Asynchronous single action
      *
      * @opensearch.internal
      */
-    class AsyncSingleAction {
+    class AsyncSingleAction extends RetryableAction {
 
-        private final ActionListener<Response> listener;
+        private ActionListener<Response> listener;
         private final Request request;
         private ClusterStateObserver observer;
         private final long startTime;
         private final Task task;
+        private boolean localRequest;
 
         AsyncSingleAction(Task task, Request request, ActionListener<Response> listener) {
+            super(logger, threadPool, TimeValue.timeValueMillis(10), request.clusterManagerNodeTimeout, listener);
             this.task = task;
             this.request = request;
-            this.listener = listener;
             this.startTime = threadPool.relativeTimeInMillis();
+            localRequest = !request.remoteRequest;
+        }
+
+        @Override
+        public void tryAction(ActionListener retryListener) {
+            ClusterState state = clusterService.state();
+            logger.trace("starting processing request [{}] with cluster state version [{}]", request, state.version());
+            this.listener = retryListener;
+            doStart(state);
+        }
+
+        @Override
+        public boolean shouldRetry(Exception e) {
+            if (localRequest) {
```

Comment:
how is remote node retrying `MasterTaskThrottlingException` now ?

Reply:
When request would be made to remote master node, we will set `remoteRequest` flag in it and send it to master. 

For throttling exception, master will not perform the retry on it based on this check and let the exception flow to the data node and data node will perform the retry. 

Since same code block is getting run for both remote/local master we need this segregation.

Reply:
As RetryableAction is triggering final listener after all the retries have exhausted, that explains why there is a need to differentiate local vs remote call? Can we differentiate by checking the transport request sourceNode instead of changing the request object?

Reply:
Sure made changes to rely on `remoteAddress` of the request, which will be null for local request for remote address it will have remote node's transport address. 

Removed the new filed from the request.